content/blog/2021-07-06-hyperparam-tuning.md

```diff
@@ -0,0 +1,280 @@
+`tuning.py`. Inside of the file, add the following code.
+
+```python
+import itertools
+import subprocess
+import random
+
+# Automated grid search experiments
+param_grid = {
+    "n_est_values": [250, 300, 350, 400, 450, 500],
+    "min_split_values": [8, 16, 32, 64, 128, 256]
+}
+param_combos = [dict(zip(param_grid.keys(), v)) for v in itertools.product(*param_grid.values())]
+
+for params in param_combos:
+  subprocess.run(["dvc", "exp", "run", "--queue", "-S", f"train.n_est={params['n_est_values']}", "-S", f"train.min_split={params['min_split_values']}"])
+```
+
+This is a simple grid search. We have two hyperparameters we want to tune:
+`n_est` and `min_split`. So we have arrays with a few values in them to mimic
+the exhaustive search a grid search can handle. Then we loop through the values
+and create queued experiments for them using `subprocess`.
+
+You can run this script now and generate your queue with this command.
+
+`python src/tuning.py`
+
+You'll see some outputs in the terminal telling you that your experiments have
+been queued. Then you can run them all with the following command.
+
+```dvc
+dvc exp run --run-all
+```
+
+This will run every experiment that has been queued. Once all of those have run,
+take a look at your metrics for each experiment.
+
+```dvc
+dvc exp show
+```
+
+Your table should look similar to this when you run `dvc exp show`.
+
+```bash
+table
+```
+
+Now you can see how your precision changed with each hyperparameter value
+update. This is a quick implementation of grid search in DVC. You could read the
+hyperparameter values from a different file or data source or make this tuning
+script as fancy as you like. The main thing you need is the
+`dvc exp run --queue -set-param <param>` command to execute when you add new
+values.
+
+### Random Search
+
+Another commonly used method for tuning hyperparameters is random search. This
+takes random values for hyperparameters and builds the model with them. It's not
+an exhaustive search like grid search and it typically outperforms grid search,
+especially when a small subset of the hyperparameters produce the biggest
+difference in your model.
+
+We're going to add an example of random search to the `tuning.py` file we
+created for grid search. This will add queued experiments with the randomly
+selected hyperparameter values. Add the following code to `tuning.py` below the
+grid search implementation.
+
+```python
+# Automated random search experiments
+num_exps = 10
+
+for _ in range(num_exps):
+    params = {
+        "rand_n_est_value": random.randint(250, 500),
+        "rand_min_split_value": random.choice([8, 16, 32, 64, 128, 256])
+    }
+    subprocess.run(["dvc", "exp", "run", "--queue", "-S",
+                   f"train.n_est={params['rand_n_est_value']}", "-S", f"train.min_split={params['rand_min_split_value']}"])
+```
+
+This random search could be far more complex with a Gaussian or Bayesian
```

Comment:
The mention of Bayesian/Gaussian might belong better in the conclusion since it's not really a random search at that point. It's about pointing out more complex tuning strategies, right?

In that case, "Bayesian optimization" is a more common phrase, and it sometimes relies on "Gaussian processes" for its surrogate model (maybe just leave out the Gaussian reference). Readers with stats backgrounds might find it odd to read about random search with "Gaussian or Bayesian distribution." Hopefully I'm not misunderstanding your intent here.

Reply:
You understood it fine. I was just using the wrong terms. 🙁 I just used Bayesian optimization and left out the other references to Bayesian/Gaussian distributions completely. They don't really add much here.